mlir/lib/Conversion/ReconcileUnrealizedCasts/ReconcileUnrealizedCasts.cpp

```diff
@@ -22,113 +20,87 @@ using namespace mlir;
-      // `{A -> B, B -> C}`, `{A -> B}`), then the cast chain is incomplete.
-
-      bool isCycle = current.getResultTypes() == op.getInputs().getTypes();
-
-      if (isLive && !isCycle)
-        return rewriter.notifyMatchFailure(op,
-                                           "live unrealized conversion cast");
-
-      bool isExitNode = users.empty() || isLive;
-
-      if (isExitNode) {
-        exitNodes.insert(current);
-      } else {
-        intermediateNodes.insert(current);
-      }
-    }
-
-    // Replace the sink nodes with the root input values
-    for (UnrealizedConversionCastOp exitNode : exitNodes)
-      rewriter.replaceOp(exitNode, op.getInputs());
-
-    // Erase all the other casts belonging to the DAG
-    for (UnrealizedConversionCastOp castOp : intermediateNodes)
-      rewriter.eraseOp(castOp);
-
-    return success();
-  }
-};
-
 /// Pass to simplify and eliminate unrealized conversion casts.
+///
+/// This pass processes unrealized_conversion_cast ops in a worklist-driven
+/// fashion. For each matched cast op, if the chain of input casts eventually
+/// reaches a cast op where the input types match the output types of the
+/// matched op, replace the matched op with the inputs.
+///
+/// Example:
+/// %1 = unrealized_conversion_cast %0 : !A to !B
+/// %2 = unrealized_conversion_cast %1 : !B to !C
+/// %3 = unrealized_conversion_cast %2 : !C to !A
+///
+/// In the above example, %0 can be used instead of %3 and all cast ops are
+/// folded away.
 struct ReconcileUnrealizedCasts
     : public impl::ReconcileUnrealizedCastsBase<ReconcileUnrealizedCasts> {
   ReconcileUnrealizedCasts() = default;
 
   void runOnOperation() override {
-    RewritePatternSet patterns(&getContext());
-    populateReconcileUnrealizedCastsPatterns(patterns);
-    ConversionTarget target(getContext());
-    target.addIllegalOp<UnrealizedConversionCastOp>();
-    if (failed(applyPartialConversion(getOperation(), target,
-                                      std::move(patterns))))
-      signalPassFailure();
+    // Gather all unrealized_conversion_cast ops.
+    SetVector<UnrealizedConversionCastOp> worklist;
+    getOperation()->walk(
+        [&](UnrealizedConversionCastOp castOp) { worklist.insert(castOp); });
+
+    // Helper function that adds all operands to the worklist that are an
+    // unrealized_conversion_cast op result.
+    auto enqueueOperands = [&](UnrealizedConversionCastOp castOp) {
+      for (Value v : castOp.getInputs())
+        if (auto castOp = v.getDefiningOp<UnrealizedConversionCastOp>())
+          worklist.insert(castOp);
+    };
+
+    // Helper function that return the unrealized_conversion_cast op that
+    // defines all inputs of the given op (in the same order). Return "nullptr"
+    // if there is no such op.
+    auto getInputCast =
+        [](UnrealizedConversionCastOp castOp) -> UnrealizedConversionCastOp {
+      if (castOp.getInputs().empty())
```

Comment:
Can a cast really have no inputs? That's a big TIL if that's true.

Reply:
Yes. E.g., this can happen when a block is converted and a block argument is dropped. (We have to provide replacement values from the new block arguments to the original bbarg types.)

Reply:
This goes a bit outside the scope of this pass and I may be wrong, but it seems to me that invalid IR would be coming into the reconciliation pass. Following your example, if the cast is there and its operand is a block operand to be removed, then the use of that argument inside the cast should be replaced by some other SSA value. Otherwise, what would the cast operate on? (I know unrealized casts are "virtual", but still, it would conceptually generate a result from nothing)

Reply:
That replacement SSA value is typically an "argument materialization". You can configure the type converter such that it builds an op that gives you a replacement value "out of thin air". E.g., this is happening in [here](https://github.com/llvm/llvm-project/blob/edabb5c12ed9312c3366286a69e881f625102a85/mlir/test/Transforms/test-legalize-type-conversion.mlir#L15).

This should also work for `unrealized_conversion_cast`s, which are just a generic variant of (argument) materializations. (However, such `unrealized_conversion_cast`s cannot be lowered any further and you are "stuck".)

Some context: I am trying to decouple argument/source/target materializations from the dialect conversion. As part of a [larger effort](https://discourse.llvm.org/t/rfc-a-new-one-shot-dialect-conversion-driver/79083/33) to simplify the current dialect conversion and eventually provide a new driver. So I'd like to insert the `unrealized_conversion_cast` without an input and let the user replace it with a proper argument materialization op after the dialect conversion. (Basically what Alex suggested in the RFC.)

Reply:
Should we maybe add a test for this as well? I don't think there is a test exercising this yet

Reply:
Added a test case with a cast op that has no inputs. (The pass does not do anything with it.)